experimental/oak_baremetal_channel/SPEC.MD

```diff
@@ -0,0 +1,296 @@
+
+##### invocation_id, u32, little endian
+
+MUST match the invocation_id of the relevant Request Message
+
+##### status_code, u32, little endian
+
+response status as a
+[gRPC status code](https://grpc.github.io/grpc/core/md_doc_statuscodes.html)
+
+##### padding, 4 bytes
+
+reserved to maintain 8 byte alignment. MUST be unset and ignored
+
+##### body, variable length byte array
+
+For responses with an "Ok" status code the body MUST be the byte encoded method
+return value. Else the body MUST be a utf-8 encoded developer-facing error
+message.
+
+#### Sending Messages
+
+To send a message, it MUST first be encoded into frames. The resulting frame set
+MUST be sent as an uninterrupted sequence.
+
+#### Receiving Messages
+
+To receive a message the recipent incrementally reads frames from the
+communication channel. The recipent MAY start parsing a message before all of
+its frames have been received.
+
+#### Encoding Messages into Frames
+
+The byte representation of a message MUST be chunked into a set of frame bodies.
+The last frame body of the resulting set MAY be shorter than 63,936 bytes (the
+maximum size of a frame body). Any preceding frame bodies MUST be exactly 63,936
+bytes in length.
+
+A frame MUST then be formed from each frame body.
+
+Messages with a length of 63,936 bytes or less MUST be encoded as a single
+frame. In this case the body of the resulting frame contains the entire message.
+
+##### Examples for Illustration
+
+#### Decoding Frames into Messages
+
+Messages are parsed from Frames by appending the frame bodies in the order they
+are received.
+
+### Frames
+
+Messages are sent and received as frames. Each message is fragmented into a set
+of one or more frames.
+
+#### Frame Fields
+
+A byte encoded frame MUST consist of the following fields:
+
+```
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                            padding                            |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|             length            |             flags             |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                                                               |
++                         body bytes...                         +
+|                                                               |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ...
+```
+
+##### padding, 4 bytes
+
+for future use
+
+##### length, u16, little endian
+
+Length of the entire frame, including this field. The length of a frame MUST NOT
+exceed 64000 bytes
+
+##### flags, two bytes, little endian
+
+###### flags[0]
+
+```
+- 7 (MSG): unused
+- 6: unused
+- 5: unused
+- 4: unused
+- 3: unused
+- 2: unused
+- 1: end flag. MUST be set in the last frame of a message
+- 0 (LSB): start flag. MUST be set in the first frame of a message
```

Comment:
In a single byte, the LSB is on the "right", so I think this is the wrong way around?

e.g. in `00000001` , the 1 is the LSB (and endianness is irrelevant for this, it's just how bytes and numbers are normally read)

So I think the fuchsia page does it correctly, i.e. the MSB (i.e. bit #7) is the "first" one.

Please others double check this though, I don't trust myself to always get this right :)

Reply:
Hehe you really shouldn't rely on me out of all people to get this right. :p 

But my two cents nonetheless: It seems to me that the convention is to start counting flags from the LSB on the right. 

I think because A) Fuchsia does this, where they list the bits in descending order from MSB to LSB (but in Flags[0] where two flags are defined and 6 undefined the defined flags start the LSB). 
B) Likewise the same appears to be the case here: https://blog.podkalicki.com/bit-level-operations-bit-flags-and-bit-masks/

Reply:
Ah that's a good point, when implementing flag masks with bit shifts, we indeed usually start counting from the right. So yes, I think your current version makes sense. Anyways if we change our mind later on for whatever reason (e.g. if the implementation is more elegant one way or the other) we can always flip things around. Although now it looks a bit off that the first flag is in the right hand side of the first flag byte, which is like in between the two bytes. Anyways let's go ahead with this for now.

Reply:
yeah I know, it's not pretty with the two bytes. But it's what fuchsia does, so at least we're not alone :p